### PR TITLE
Enforce stronger head state operations

### PR DESCRIPTION
**Pull request by prestonvanloon**
- Don't save a head state block root if no state exists with that root
- Don't delete a head state from the database

~Note: breaks `--initial-sync-cache-state` which saves head block root but doesn't save state in db.~ Fixed in #4219

### CODE REVIEW

**Comment by prestonvanloon** [comment]
A few tests left to resolve. Will pick up tmrw

**Comment by prestonvanloon** [comment]
Note for reviewers: review the code in the db/kv package. Everything else is test resolution to adhere to the new policy.

File: beacon-chain/db/kv/state.go [hunk 1]
```diff
@@ -116,9 +116,12 @@ func (k *Store) DeleteState(ctx context.Context, blockRoot [32]byte) error {
...
 
-		// Safe guard against deleting genesis or finalized state.
-		if bytes.Equal(blockRoot[:], checkpoint.Root) || bytes.Equal(blockRoot[:], genesisBlockRoot) {
-			return errors.New("could not delete genesis or finalized state")
+		bkt = tx.Bucket(blocksBucket)
+		headBlkRoot := bkt.Get(headBlockRootKey)
+
+		// Safe guard against deleting genesis, finalized, head state.
+		if bytes.Equal(blockRoot[:], checkpoint.Root) || bytes.Equal(blockRoot[:], genesisBlockRoot) || bytes.Equal(blockRoot[:], headBlkRoot) {
+			return errors.New("could not delete genesis, finalized, or head state")
```
Comment (rauljordan) [inline beacon-chain/db/kv/state.go]:
```suggestion
			return errors.New("cannot delete genesis, finalized, or head state")
```

File: beacon-chain/db/kv/state.go [hunk 2]
```diff
@@ -144,10 +147,13 @@ func (k *Store) DeleteStates(ctx context.Context, blockRoots [][32]byte) error {
...
 
+		bkt = tx.Bucket(blocksBucket)
+		headBlkRoot := bkt.Get(headBlockRootKey)
+
 		for _, blockRoot := range blockRoots {
-			// Safe guard against deleting genesis or finalized state.
-			if bytes.Equal(blockRoot[:], checkpoint.Root) || bytes.Equal(blockRoot[:], genesisBlockRoot) {
-				return errors.New("could not delete genesis or finalized state")
+			// Safe guard against deleting genesis, finalized, or head state.
+			if bytes.Equal(blockRoot[:], checkpoint.Root) || bytes.Equal(blockRoot[:], genesisBlockRoot) || bytes.Equal(blockRoot[:], headBlkRoot) {
```
Comment (rauljordan) [inline beacon-chain/db/kv/state.go]: Duplicated logic here, can it be its own helper func?

File: beacon-chain/db/kv/state_test.go [hunk 3]
```diff
@@ -214,8 +201,43 @@ func TestStore_DeleteFinalizedState(t *testing.T) {
...
+		t.Fatal(err)
+	}
+	headState := &pb.BeaconState{Slot: 100}
+	if err := db.SaveState(ctx, headState, headBlockRoot); err != nil {
+		t.Fatal(err)
+	}
+	if err := db.SaveHeadBlockRoot(ctx, headBlockRoot); err != nil {
+		t.Fatal(err)
+	}
+	wantedErr := "could not delete genesis, finalized, or head state"
```
Comment (rauljordan) [inline beacon-chain/db/kv/state_test.go]:
```suggestion
	wantedErr := "cannot delete genesis, finalized, or head state"
```

File: beacon-chain/db/kv/state.go (hunk 2 above)
Comment (nisdas) [reply beacon-chain/db/kv/state.go]: +1

Comment (prestonvanloon) [reply beacon-chain/db/kv/state.go]: Looking at this again, it's a bit difficult to refactor this. I suppose I could write a method like
```go
func ensureStateRoots(blockRoot, finalizedCheckpointRoot, genesisBlockRoot, headBlockRoot [32]byte) bool {
  return bytes.Equal(blockRoot[:], checkpoint.Root) || bytes.Equal(blockRoot[:], genesisBlockRoot) || bytes.Equal(blockRoot[:], headBlkRoot) 
}
```

I don't know if this adds much value though. What do you think?

Comment (nisdas) [reply beacon-chain/db/kv/state.go]: imo, in that case lets just keep it as it is, if there isnt much benefit to changing it

**Pull request merged**